### DISCUSSION TITLE
Composing two images on top so that the second one is always readable

### DISCUSSION THREAD

**Discussion by HydrodynamicSynchrotron** [ImageMagick/ImageMagick#4574] [Help] [answered]
Hello,

this is a stupid summary, here's what I'm trying to do:

I have an arbitrary source image, a picture. Some very dark, some very bright. I have a second image, a watermark. Black text on transparent background. I want to compose the two images on top of each other.

Problem: Dark text on a dark picture is almost unreadable. Hence I want to change the composition so that it's not just putting them on top of each other, but for each pixel: If there's no alpha set, put black if the background was bright (L > 150) and put white if the background was dark (L <= 150).

I have tried fiddling with the composition, mathematical composition and fx operator, but cannot seem to get it to work. Could someone please help me? Thank you.

**Comment by fmw42** [comment]
Please post example images. What is your ImageMagick version and platform/OS?

**Reply by HydrodynamicSynchrotron** [reply]
Hello, sorry forgot this: I'm using ImageMagick 6.9.11-60 Q16 x86_64 2021-01-25 on Linux. The attached example image is CC-0 and on top of it the text should be overlayed, but so that it's readable (i.e., where the background is white, the text should be black/dark, where the background is black, the text should appear as white/light). Thank you.

![image](https://user-images.githubusercontent.com/8970887/145641321-28203a7e-23fe-41af-b45e-31ff037759b7.jpg)
![text](https://user-images.githubusercontent.com/8970887/145641324-2f704357-e540-43b1-9018-a40998066198.png)

**Comment by fmw42** [comment, marked as answer]
This may not be the most efficient, but it works for me.

```
convert 145641321-28203a7e-23fe-41af-b45e-31ff037759b7.jpg +write mpr:img \
\( 145641324-2f704357-e540-43b1-9018-a40998066198.png -alpha extract +write mpr:mask \) \
-compose multiply -composite -negate \
mpr:mask -compose multiply -composite \
mpr:img +swap mpr:mask -compose over -composite \
result.jpg
```

![result](https://user-images.githubusercontent.com/13458218/145648907-8ae75857-1c56-4f91-8282-f17780f58908.jpg)

**Comment by HydrodynamicSynchrotron** [comment]
That is AMAZING, exactly what I was looking for. A million thanks!
